File: AOJ/0033.cpp
#include <iostream>
#include <vector>
using namespace std;

int N;
int balls[15];

int solve() {
    vector<int> A, B;
    for(int i = 0; i < 10; i++) {
        if(i == 0) {
            A.push_back(balls[i]);
        } else {
            if(B.size() == 0) {
                if(A[A.size()-1] > balls[i]) {
                    B.push_back(balls[i]);
                } else {
                    A.push_back(balls[i]);
                }
            } else {
                if(A[A.size()-1] < balls[i] && B[B.size()-1] < balls[i]) {
                    if(A[A.size()-1] > B[B.size()-1]) {
                        A.push_back(balls[i]);
                    } else {
                        B.push_back(balls[i]);
                    }
                } else if(A[A.size()-1] < balls[i] && B[B.size()-1] > balls[i]) {
                    A.push_back(balls[i]);
                } else if(A[A.size()-1] > balls[i] && B[B.size()-1] < balls[i]) {
                    B.push_back(balls[i]);
                } else {
                    cout << "NO" << endl;
                    return 0;
                }
            }
        }
    }

    cout << "YES" << endl;
    return 0;
}


int main() {
    cin >> N;
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < 10; j++) {
            cin >> balls[j];
        }
        solve();
    }

    return 0;
}

File: AOJ/0033.java
import java.util.Scanner;

/**
 * Created by sotetsuk on 2015/07/31.
 */
public class AOJ0033 {
    static final int K = 10;
    static int N;
    static int[] balls = new int[10];
    static Scanner sc = new Scanner(System.in);
    static int l, r;

    public static void read() {
        for(int i = 0; i < K; i++) {
            balls[i] = sc.nextInt();
        }
    }

    public static boolean dfs(int i) {
        // stop condition
        if(i == K) return true;

        // recursive
        boolean flag_l = false; boolean flag_r = false;
        if(l == -1 || balls[l] < balls[i]) {
            int tmp_l = l; l = i; flag_l = dfs(i + 1); l = tmp_l;
        }
        if(r == -1 || balls[r] < balls[i]) {
            int tmp_r = r; r = i; flag_r = dfs(i + 1); r = tmp_r;
        }
        return flag_l || flag_r;
    }

    public static void solve() {
        l = -1; r = -1;
        String ans = dfs(0) ? "YES" : "NO";
        System.out.println(ans);
    }

    public static void main(String[] args) {
        N = sc.nextInt();
        for(int i = 0; i < N; i++) {
            read();
            solve();
        }

    }
}
